pinakes/main/approval/models.py

```diff
@@ -184,9 +185,124 @@ class Meta:
+
+    def _internal_sequence_plus(self, delta):
+        if delta == math.inf:
+            return self._internal_sequence_extend_from_last()
+
+        query = Workflow.objects.filter(
+            internal_sequence__gt=self.internal_sequence
+        )
+        sequence_range = self._internal_sequence_range(query, delta)
+
+        if len(sequence_range) == 0:
+            return self._internal_sequence_extend_from_last()
+
+        if len(sequence_range) == 1:
+            return sequence_range[0] + Workflow.INTERNAL_INTERVAL
+
+        if sequence_range[1] - sequence_range[0] < 2:
+            self._rebalance_internal_sequences()
+            return self._internal_sequence_plus(delta)
+
+        return int((sequence_range[0] + sequence_range[1]) / 2)
+
+    def _internal_sequence_minus(self, delta):
+        if delta == math.inf:
+            return self._internal_sequence_before_first()
+
+        query = Workflow.objects.filter(
+            internal_sequence__lt=self.internal_sequence
+        ).order_by("-internal_sequence")
+        sequence_range = self._internal_sequence_range(query, delta)
+        if len(sequence_range) == 0:
+            return self._internal_sequence_before_first()
+
+        if len(sequence_range) == 1:
+            return self._half_first_internal_sequence(sequence_range[0])
+
+        if sequence_range[0] - sequence_range[1] < 2:
+            self._rebalance_internal_sequences()
+            return self._internal_sequence_minus(delta)
+
+        return int((sequence_range[0] + sequence_range[1]) / 2)
+
+    def _internal_sequence_range(self, query, delta):
+        return query.values_list("internal_sequence", flat=True)[
+            (delta - 1) : (delta + 1)
+        ]
+
+    def _internal_sequence_extend_from_last(self):
+        last_sequence = (
+            Workflow.objects.filter(tenant=self.tenant)
+            .values_list("internal_sequence", flat=True)
+            .last()
+        )
+        return last_sequence + Workflow.INTERNAL_INTERVAL
+
+    def _internal_sequence_before_first(self):
+        first_sequence = (
+            Workflow.objects.filter(tenant=self.tenant)
+            .values_list("internal_sequence", flat=True)
+            .first()
+        )
+        return self._half_first_internal_sequence(first_sequence)
+
+    def _half_first_internal_sequence(self, first_sequence):
+        if first_sequence < 2:
+            self._rebalance_internal_sequences()
+            return int(Workflow.INTERNAL_INTERVAL / 2)
+        return int(first_sequence / 2)
+
+    def _rebalance_internal_sequences(self):
+        workflows = Workflow.objects.filter(tenant=self.tenant)
+        new_sequence = 0
+        for workflow in workflows:
+            new_sequence = new_sequence + Workflow.INTERNAL_INTERVAL
+            workflow.internal_sequence = new_sequence
+
+        Workflow.objects.bulk_update(workflows, ["internal_sequence"], 100)
```

Comment:
This hardcoded number plays trivial factor. The value does not matter much and almost never needs adjustment.

Reply:
This trivial factor doesn't seem to be local to the current query. Bulk updates may be used in other parts of the project. So hard coding it here would make less straightforward changing it in future if necessary.

Reply:
https://docs.djangoproject.com/en/4.0/ref/models/querysets/ does not indicate the batch_size argument is global. I believe it is local. The default value is None, meaning all qualified workflows will be updated in one shot. We can use default value as well because in reality we don't expect the customer will add a lot of workflows. I choose 100 because I saw examples using this value.

Reply:
It seems you understood my comment in a different way. The `100` number is not exclusive to that code. In case any other bulk_update operations appear, more hard coded numbers appear as a result. In case this number is considered sane default for the batch size, changing it across the code would require additional refactoring and searching for those magic numbers.

Reply:
I think I agree with @cutwater here, the `100` seems like a fine default. I know it's currently only used in this one place, but it does seem to make sense to me to change it to a constant that is maybe more descriptive and easier to change later. Something like `DEFAULT_BULK_UPDATE_VALUE`?